### PR TITLE
Convert hash API to use MP_UNARY_OP_HASH instead of ad-hoc function

### PR DESCRIPTION
**Pull request by dpgeorge**
Hashing is now done using mp_unary_op function with MP_UNARY_OP_HASH as
the operator argument.  Hashing for int, str and bytes still go via
fast-path in mp_unary_op since they are the most common objects which
need to be hashed.

This lead to quite a bit of code cleanup, and should be more efficient
if anything.  It saves 176 bytes code space on Thumb2, and 360 bytes on
x86.

The only loss is that the error message "unhashable type" is now the
more generic "unsupported type for **hash**".

To address issue #1246.

@pfalcon it turned out that I could save a significant number of bytes and at the same time (hopefully) cover all your points.  Please take a critical look!

### CODE REVIEW

File: py/objtype.c [hunk 1]
```diff
@@ -355,8 +356,28 @@ STATIC mp_obj_t instance_unary_op(mp_uint_t op, mp_obj_t self_in) {
     if (member[0] == MP_OBJ_SENTINEL) {
         return mp_unary_op(op, self->subobj[0]);
     } else if (member[0] != MP_OBJ_NULL) {
-        return mp_call_function_1(member[0], self_in);
+        mp_obj_t val = mp_call_function_1(member[0], self_in);
+        // __hash__ must return a small int
+        if (op == MP_UNARY_OP_HASH) {
+            val = MP_OBJ_NEW_SMALL_INT(mp_obj_get_int(val));
```
Comment (dpgeorge) [inline py/objtype.c]: Actually, this should be mp_obj_get_int_truncated.

Comment (pfalcon) [inline py/objtype.c]: +1

**Comment by pfalcon** [comment]
Well, I imagined more localized refactor, where mp_obj_hash() stays the same for simple types and dispatches to ->unary_op for more complex ones, but if that's what allowed to get more code size savings, looks good, thanks for going for that!

**Comment by dpgeorge** [comment]
It was natural to put small-int hash in unary op with all the other small-int unary ops there already.  And then only str hashing was left as the most common one needing a fast path (and actually needed a better fast path than existing in mp_obj_hash, as per old comment in objstr.c).  So mp_obj_hash became reduntant and everything worked out for the best.

**Comment by dpgeorge** [comment]
Merged in c2a4e4effc81d8ab21bb014e34355643e5ca0da2.

@stinos you can now hopefully fix your original problem with hashing of native types.

**Pull request closed without merging**